server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java

```diff
@@ -0,0 +1,138 @@
+            final TDigestState approximatedDeviationsSketch = new TDigestState(valuesSketch.compression());
+
+            valuesSketch.centroids().forEach(centroid -> {
+                final double deviation = Math.abs(approximateMedian - centroid.mean());
+                approximatedDeviationsSketch.add(deviation, centroid.count());
+            });
+
+            return approximatedDeviationsSketch.quantile(0.5);
+        }
+    }
+
+    protected final TDigestState valuesSketch;
+    protected final double medianAbsoluteDeviation;
+
+    public InternalMedianAbsoluteDeviation(String name,
+                                           List<PipelineAggregator> pipelineAggregators,
+                                           Map<String, Object> metaData,
+                                           DocValueFormat format,
+                                           TDigestState valuesSketch) {
+
+        super(name, pipelineAggregators, metaData);
+        this.format = Objects.requireNonNull(format);
+        this.valuesSketch = Objects.requireNonNull(valuesSketch);
+
+        this.medianAbsoluteDeviation = computeMedianAbsoluteDeviation(this.valuesSketch);
+    }
+
+    public InternalMedianAbsoluteDeviation(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        valuesSketch = TDigestState.read(in);
+        medianAbsoluteDeviation = in.readDouble();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        TDigestState.write(valuesSketch, out);
+        out.writeDouble(medianAbsoluteDeviation);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        final TDigestState valueMerged = new TDigestState(valuesSketch.compression());
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalMedianAbsoluteDeviation madAggregation = (InternalMedianAbsoluteDeviation) aggregation;
+            valueMerged.add(madAggregation.valuesSketch);
+        }
+
+        return new InternalMedianAbsoluteDeviation(name, pipelineAggregators(), metaData, format, valueMerged);
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        final boolean anyResults = valuesSketch.size() > 0;
+        final Double mad = anyResults
+            ? getMedianAbsoluteDeviation()
+            : null;
+
+        builder.field(CommonFields.VALUE.getPreferredName(), mad);
+        if (format != DocValueFormat.RAW && anyResults) {
+            builder.field(CommonFields.VALUE_AS_STRING.getPreferredName(), format.format(mad).toString());
+        }
+
+        return builder;
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(valuesSketch, medianAbsoluteDeviation);
```

Comment:
Not sure if the MAD should be included here and equals 🤔 

I believe equal values sketches implies equal MAD

Reply:
Agreed